dbos/workflow.go

```diff
@@ -607,37 +612,68 @@ func (c *dbosContext) RunAsWorkflow(_ DBOSContext, fn WorkflowFunc, input any, o
 		stepID:     -1,
 	}
 
-	// Run the function in a goroutine
 	workflowCtx := WithValue(c, workflowStateKey, wfState)
+
+	// If the workflow has a deadline, set it in the context. We use what was returned by InsertWorkflowStatus
+	var stopFunc func() bool
+	cancelFuncCompleted := make(chan struct{})
+	if !insertStatusResult.workflowDeadline.IsZero() {
+		workflowCtx, _ = WithTimeout(workflowCtx, time.Until(insertStatusResult.workflowDeadline))
+		// Register a cancel function that cancels the workflow in the DB as soon as the context is cancelled
+		dbosCancelFunction := func() {
+			c.logger.Info("Cancelling workflow", "workflow_id", workflowID)
+			err = c.systemDB.CancelWorkflow(uncancellableCtx, workflowID)
+			if err != nil {
+				c.logger.Error("Failed to cancel workflow", "error", err)
+			}
+			close(cancelFuncCompleted)
+		}
+		stopFunc = context.AfterFunc(workflowCtx, dbosCancelFunction)
+	}
+
+	// Run the function in a goroutine
 	c.workflowsWg.Add(1)
 	go func() {
 		defer c.workflowsWg.Done()
 		result, err := fn(workflowCtx, input)
 		status := WorkflowStatusSuccess
+
+		// If an error occurred, set the status to error
 		if err != nil {
 			status = WorkflowStatusError
 		}
-		recordErr := c.systemDB.UpdateWorkflowOutcome(c.ctx, updateWorkflowOutcomeDBInput{
+
+		// If the afterFunc has started, the workflow was cancelled and the status should be set to cancelled
+		if stopFunc != nil && !stopFunc() {
+			// Wait for the cancel function to complete
+			// Note this must happen before we write on the outcome channel (and signal the handler's GetResult)
+			<-cancelFuncCompleted
+			// Set the status to cancelled and move on so we still record the outcome in the DB
+			status = WorkflowStatusCancelled
+		}
```

Comment:
For whoever is interested in how this stuff works: https://pkg.go.dev/context#AfterFunc